src/SwedbankPay.Sdk/SwedbankPayHttpClient.cs

```diff
@@ -0,0 +1,173 @@
+using Newtonsoft.Json.Serialization;
+
+using SwedbankPay.Sdk.Exceptions;
+
+namespace SwedbankPay.Sdk
+{
+    internal class SwedbankPayHttpClient
+    {
+        private readonly HttpClient client;
+        private readonly ILogger logger;
+
+
+        internal SwedbankPayHttpClient(HttpClient client, ILogger logger)
+        {
+            this.client = client;
+            this.logger = logger;
+        }
+
+
+        internal async Task<string> GetRaw(string url)
+        {
+            var msg = new HttpRequestMessage(HttpMethod.Get, url);
+            msg.Headers.Add("Accept", "application/json");
+
+            HttpResponseMessage response;
+            try
+            {
+                response = await this.client.SendAsync(msg);
+            }
+            catch (HttpRequestException e)
+            {
+                throw new BadRequestException(e);
+            }
+            catch (TaskCanceledException te)
+            {
+                throw new ApiTimeOutException(te);
+            }
+
+            if (response.IsSuccessStatusCode)
+            {
+                var res = await response.Content.ReadAsStringAsync();
+                var res2 = JToken.Parse(res).ToString(Formatting.Indented);
+                this.logger.LogInformation(res2);
+                return res2;
+            }
+
+            return null;
+        }
+
+
+        internal async Task<TResponse> HttpGet<TResponse>(string url, Func<ProblemsContainer, Exception> onError)
+            where TResponse : new()
+        {
+            return await SendHttpRequestAndProcessHttpResponse<TResponse>(HttpMethod.Get, url, onError);
+        }
+
+
+        internal async Task<TResponse> HttpPatch
+            <TPayLoad, TResponse>(string url, Func<ProblemsContainer, Exception> onError, TPayLoad payload)
+            where TResponse : new()
+        {
+            return await SendHttpRequestAndProcessHttpResponse<TResponse>(new HttpMethod("PATCH"), url, onError, payload);
+        }
+
+
+        internal async Task<TResponse> HttpPost
+            <TPayLoad, TResponse>(string url, Func<ProblemsContainer, Exception> onError, TPayLoad payload)
+            where TResponse : new()
+        {
+            return await SendHttpRequestAndProcessHttpResponse<TResponse>(HttpMethod.Post, url, onError, payload);
+        }
+
+
+        internal async Task<TResponse> SendHttpRequestAndProcessHttpResponse
+            <TResponse>(HttpMethod httpMethod, string url, Func<ProblemsContainer, Exception> onError, object payload = null)
+            where TResponse : new()
+        {
+            var requestMessage = new HttpRequestMessage(httpMethod, url);
+
+            return await SendHttpRequestAndProcessHttpResponse<TResponse>(requestMessage, onError, payload);
+        }
+
+
+        internal async Task<TResponse> SendHttpRequestAndProcessHttpResponse
+            <TResponse>(HttpRequestMessage requestMessage, Func<ProblemsContainer, Exception> onError, object payload = null)
```

Comment:
Consider taking in only the requestMessage and not the payload as well. It's really messy having to send around two arguments that are only going to get merged further down the road anyway.